rfcs/0031-devices-api.md

```diff
@@ -0,0 +1,363 @@
+        scope.stack_tcode,
+        ConstInt32(arg_stack_begin),
+        ConstInt32(arg_stack_begin + op->args.size() - 1),
+        resource_handle
+    }
+);
+```
+
+## Device Discovery
+Initially, devices will be defined by Target name or external compiler name. This means if you mark an operator as needing an external `woofles` compiler it would result in a devices struct such as:
+
+```c
+struct tvmgen_my_model_devices {
+    tvm_device_woofles_t* woofles
+};
+```
+
+Which would be passed down to the relevant operators via the executor. This applies similarly to `Target` defined devices.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* Current limitations with `Target` and external compilers mean that only one of each name can occur at once using this system, this could equally be in future work.
+* The initial assumption is that each operator will be mapped to a single device, this design choice means that fusion across devices will not be possible.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could leverage more code generation to generate device structures. It is the authors belief that being able to write small self-contained platform implementations will be easier to understand for both users and developers of TVM.
+
+Another route to take is to treat RTOSes as entirely separate from TVM, requiring them to fully configure resources before passing in the `void*`. This removes TVMs ability to add hooks for resource management such as `open` and `close` which could be used to enable/disable entire pieces of circuitry between operators.
+
+# Prior art
+[prior-art]: #prior-art
+* Uses the existing `resource_handle` in the TVM code which isn't currently propagated
+* Extends the C Interface API to add support for devices
+* Resource management using `open`/`close` and `init`/`destroy` alongside opaque handles is a common pattern in C libraries
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+This RFC aims to put in place the foundation of the Device API to start abstracting the various RTOS drivers. There are other flows that have been considered as extensions to this.
+
+## Memory Copies
+Movement of memory between additional devices which may be unable to communicate directly, this could take the form of simply:
+
+```
+// Copy from/to
+int32_t TVMDeviceWooflesCopyFrom(tvm_device_t* source, void* destination);
+int32_t TVMDeviceUartAccelCopyTo(void* source, tvm_device_t* destination);
+```
+
+And be integrated into the flow as follows:
+
+```
+TVMDeviceWooflesOpen(device1);
+operator(..., device1) {
+    // some work where device1 can read from memory directly
+    // then the result is copied back
+    TVMDeviceWooflesCopyFrom(device1, &buffer);
+}
+TVMDeviceWooflesClose(device1);
+
+TVMDeviceUartAccelOpen(device2);
+operator(..., device2) 
+    TVMDeviceUartAccelCopyTo(&buffer, device2);{
```

Comment:
what if memory is needed across operator calls? does the device have to be Open'd for e.g. the whole runtime of the model when e.g. a parameter is pre-loaded and the device call is the last layer in the model?

Reply:
This is an example API for memory management in future possibilities, there's a much larger question as to how devices can specify all of the above as I'm unaware of which devices you're referring to and the specific characteristics in tandem with the model you're using.